### PR TITLE
NcpSpi/Spi-slave: Ensure OT APIs are not called from ISR context

### PR DESCRIPTION
**Pull request by abtink**
This commit changes the `spi-slave.h` platform functions by modifying
how the callbacks for a completed SPI transaction are used/invoked.

There are two callbacks for when a SPI transaction completes: The
`TransactionCompleteCallback` can be called from ISR context and it
prepares  the next transaction based on current internal `NcpSpi`
state quickly. This callback returns a `bool` to indicate to  platform
spi-slave driver if further processing is required. If `true` is
returned, the platform spi-slave driver is expected to invoke the
newly introduced callback `TransactionProcessCallback` from the same
OS context that all OpenThread APIs and callbacks are called. This
basically, divides the processing into two parts, one part from ISR
and second follow up from OpenThread context.

This change is to address an issue with current model where OpenThread
APIs (e.g., posting a tasklet) could be called from an ISR context
(which was unsafe and could possibly lead to dropping/missing a posted
tasklet).

### CODE REVIEW

**Comment by abtink** [comment]
**Note**: This is implemented on top of "Full duplex" commit. Setting it up now for code review (while testing it internally). Marking it as "work-in-progress" so we do not merge. Thanks.

**Comment by abtink** [comment]
Curious. The appvoyor builds seems to have failed, but on my setup it passed for same branch: https://ci.appveyor.com/project/abtink/openthread/build/0.1.223

**Comment by abtink** [comment]
Rebased this branch on master... It's a single commit now.

**Comment by abtink** [comment]
The change is tested and verified.

**Comment by darconeous** [comment]
So posting a tasklet is not ISR safe? Isn't that the real problem that should be fixed?

**Comment by abtink** [comment]
Yes. Tasklets are for internal thread use/API and posting one from ISR is not safe. 
We can change this but, it requires a major re-design. We have the rule that all OT APIs/callbacks should be called from the same context. There was a related issue/discussion: https://github.com/openthread/openthread/issues/1627, (copying the comment):

> For simplicity and testability, we landed on a design where the OpenThread API is assumed to be called from a single execution context. Any mechanisms required to handle concurrency are out-of-scope and left to platform-specific code around the core OpenThread library
